### PR TITLE
Decoupling ViewModels & Abstractions from Maui

### PR DESCRIPTION
**Pull request by sferhah**
1/2 Decoupling ViewModels & Abstractions from Maui (the objective is to create a separate project, platform agnostic, destined to be referenced by several types of projects: Maui, Xamarin, Avalonia, WinUi, Android native...etc.)

### CODE REVIEW

File: src/Jellyfin.Maui/Services/INavigationService.cs [hunk 1]
```diff
@@ -46,4 +42,26 @@ public interface INavigationService
...
+    /// <param name="action">he method to be executed by the dispatcher.</param>
+    Task DispatchAsync(Action action);
+
+    /// <summary>
+    /// Starts synchronization on the collection by using callback and context.
+    /// </summary>
+    /// <param name="collection">The collection that will be read or updated.</param>
+    /// <param name="context">The context or lock object that will be passed to callback. May be null.</param>
+    /// <param name="callback">The synchronization callback.</param>
+    void EnableCollectionSynchronization(IEnumerable collection, object? context, Action<IEnumerable, object, Action, bool> callback);
```
Comment (crobibero) [inline src/Jellyfin.Maui/Services/INavigationService.cs]: I would move all of these to a new `ApplicationService` as they don't really have anything to do with navigation.

File: src/Jellyfin.Maui/Services/NavigationService.cs [hunk 2]
```diff
@@ -9,23 +10,15 @@ namespace Jellyfin.Maui.Services;
 /// <inheritdoc />
 public class NavigationService : INavigationService
 {
-    // Application is initialized on startup.
-    private Application _application = null!;
```
Comment (crobibero) [inline src/Jellyfin.Maui/Services/NavigationService.cs:13]: Revert below changes and change L13 to the following.
```
private readonly Application _application = Application.Current!;
```

**Review by crobibero — CHANGES_REQUESTED** [review]

**Comment by sferhah** [comment]
I've just made another commit, tell me if it's ok for you. (I'm a git noob, I take no offense)

**Comment by sferhah** [comment]
I've also fixed a frequent crash within the ServerSelectViewModel (Servers.ReplaceRange(state.Servers)).

Why do you need to call .ConfigureAwait(false) in your viewModels ? .ConfigureAwait(false) leaves the Ui thread, any subsequent calls will need to be resynchronized with DispatchAsync.

**Comment by crobibero** [comment]
> Why do you need to call .ConfigureAwait(false) in your viewModels ? .ConfigureAwait(false) leaves the Ui thread, any subsequent calls will need to be resynchronized with DispatchAsync.

I'm a backend developer and just playing with Maui to learn- it is very possible that I have it incorrect. My thinking was to do all processing off of the UI thread and then sync when complete

File: src/Jellyfin.Maui/Services/IApplicationService.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+/// <summary>
+/// ApplicationService service interface.
+/// </summary>
+public interface IApplicationService
+{
+    /// <summary>
+    /// Schedules the provided action on the UI thread from a worker thread.
+    /// </summary>
+    /// <returns>true when the action has been dispatched successfully, otherwise false.</returns>
+    /// <param name="action">he method to be executed by the dispatcher.</param>
```
Comment (crobibero) [inline src/Jellyfin.Maui/Services/IApplicationService.cs]:
```suggestion
    /// <param name="action">The method to be executed by the dispatcher.</param>
```

File: src/Jellyfin.Maui/Services/IApplicationService.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    bool Dispatch(Action action);
+
+    /// <summary>
+    /// Schedules the provided action on the UI thread from a worker thread.
+    /// </summary>
+    /// <returns>true when the action has been dispatched successfully, otherwise false.</returns>
+    /// <param name="action">he method to be executed by the dispatcher.</param>
```
Comment (crobibero) [inline src/Jellyfin.Maui/Services/IApplicationService.cs]:
```suggestion
    /// <param name="action">The method to be executed by the dispatcher.</param>
```

File: src/Jellyfin.Maui/Services/NavigationService.cs [hunk 5]
```diff
@@ -9,17 +9,10 @@ namespace Jellyfin.Maui.Services;
...
+    private Application _application = Application.Current!;
```
Comment (crobibero) [inline src/Jellyfin.Maui/Services/NavigationService.cs:12]:
```suggestion
    private readonly Application _application = Application.Current!;
```

**Comment by crobibero** [comment]
I don't see how changing `ConfigureAwait(false)` to `ConfigureAwait(true)` fixes the thread update, my understanding is that `SafeFireAndForget()` brings the task to another thread.

**Comment by sferhah** [comment]
> I don't see how changing `ConfigureAwait(false)` to `ConfigureAwait(true)` fixes the thread update, my understanding is that `SafeFireAndForget()` brings the task to another thread.

You are right, I did not see the code signature of SafeFireAndForget (Never used that lib):

`public static void SafeFireAndForget(this ValueTask task, in Action<Exception>? onException = null, in bool continueOnCapturedContext = false)`

the default value of continueOnCapturedContext is false, I assumed it to be true like that of ConfigureAwait().

So in order for it to work, BaseContentPage<TViewModel> needs to explicitly set continueOnCapturedContext to true:

```
    protected override void OnNavigatedTo(NavigatedToEventArgs args)
    {
        ViewModel.InitializeAsync().SafeFireAndForget(continueOnCapturedContext: true);
        base.OnNavigatedTo(args);
    }
```

**Comment by sferhah** [comment]
I've implemented your last suggestion (fix typo), 

Anyway, let's keep the code as is. Not very important for now,  I'll try to implement Shell navigation for my next PR.

**Comment by crobibero** [comment]
I'm fairly certain `EnableCollectionSynchronization` is supposed to update the collection on the UI thread as well

**Review by crobibero — APPROVED** [review]

**Pull request merged**